### PR TITLE
In-memory engine: implement core local array and collect read flow statistics

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
implement core local array and collect read flow statistics
```
This PR implements CoreLocalArray (which is implemented in RocksDB) for collecting metrics where, ideally, each thread only report stats to the cpu core executing it to readuce contention. In addition, this PR collects read flow and report to grafana like below:
image

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by afeinberg — APPROVED** [review]
LGTM. @tonyxuqqi any thoughts on RangeCacheOptions?

File: components/region_cache_memory_engine/src/read.rs [hunk 1]
```diff
@@ -417,6 +435,14 @@ impl Iterator for RangeCacheIterator {
         if self.valid {
             self.find_next_visible_key(true, guard);
         }
+
+        if self.valid {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/read.rs:439]: why not merge with the self.valid above.

Comment (tonyxuqqi) [reply components/region_cache_memory_engine/src/read.rs:439]: Also the code is duplicated by many times, make it an shared inline function

**Review by tonyxuqqi — APPROVED** [review]
Overall LGTM. Some details can be polished.

File: components/server/src/common.rs [hunk 2]
```diff
@@ -701,15 +704,15 @@ impl<T: fmt::Display + Send + 'static> Stop for LazyWorker<T> {
 
 pub trait KvEngineBuilder: KvEngine {
     fn build(
-        range_cache_engine_config: Arc<VersionTrack<RangeCacheEngineConfig>>,
+        range_cache_engine_config: RangeCacheEngineOptions,
```
Comment (tonyxuqqi) [inline components/server/src/common.rs:707]: the name and type does not match.

File: components/region_cache_memory_engine/src/lib.rs [hunk 3]
```diff
@@ -99,3 +104,21 @@ impl RangeCacheEngineConfig {
         }
     }
 }
+
+pub struct RangeCacheEngineOptions {
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/lib.rs:108]: The naming of RangeCacheEngineOptions is not appropriate. Statistics should not be part of Option.

Comment (tonyxuqqi) [reply components/region_cache_memory_engine/src/lib.rs:108]: I think statistics should be just a member of RangeCacheMemoryEngine directly and initialized inside its constructor.

**Comment by tonyxuqqi** [comment]
> LGTM. @tonyxuqqi any thoughts on RangeCacheOptions?

RangeCacheOptions is not necessary and it's also impropriate to include statistics as part of Options

File: components/region_cache_memory_engine/src/read.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/read.rs:439]: after find_next_visible_key, the self.valid may not be true.

**Comment by SpadeA-Tang** [comment]
> > LGTM. @tonyxuqqi any thoughts on RangeCacheOptions?
> 
> RangeCacheOptions is not necessary and it's also impropriate to include statistics as part of Options

RocksDB put the `Statistics` in its db option. Also, we need this so we can get the statistics at the `TiKVServer::init_raw_engines`.

File: components/region_cache_memory_engine/src/read.rs (hunk 1 above)
Comment (tonyxuqqi) [reply components/region_cache_memory_engine/src/read.rs:439]: Please add a comment to mention that tself.valid can be changed. It's not straightforward.

Comment (tonyxuqqi) [reply components/region_cache_memory_engine/src/read.rs]: But we should not have duplicated code when possible

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/read.rs]: I have polished that.

**Review by LykxSassinator — APPROVED** [review]

**Pull request merged**